File: 2024/day03_Mull_It_Over/main.py
import sys
import re

def part1(file: list[str]):
    tuples = re.findall("mul\((\d+),(\d+)\)","".join(file))

    res = sum(map(lambda tuple: int(tuple[0])*int(tuple[1]), tuples))
    
    print(res)

def part2(file: list[str]):
    pattern = re.compile("do\(\)|don't\(\)|mul\((\d+),(\d+)\)")

    enable = True
    res = 0
    for mt in re.finditer(pattern, "".join(file)):
        match mt.group(), mt.groups():
            case ("do()", _):
                enable = True
            case ("don't()", _):
                enable = False
            case (_, (x, y)):
                if enable:
                    res += int(x) * int(y)
    
    print(res)


with open(sys.argv[1]+".txt", "r") as file:
    lines = [line.rstrip("\n") for line in file]

part1(lines)
part2(lines)

File: 2024/day03_Mull_It_Over/src/main.rs
use std::io::stdin;
//use regex
use regex::Regex;


fn part1(file: &String) {
    let pattern = Regex::new(r"mul\((\d{1,3}),(\d{1,3})\)").unwrap();

    let res = pattern
        .captures_iter(file)
        .map(|cap| {
            let a = cap[1].parse::<i32>().unwrap();
            let b = cap[2].parse::<i32>().unwrap();
            a * b
        })
        .sum::<i32>();
    println!("{}", res);
}

fn part2(file: &String) {
    let pattern = Regex::new(r"(do(?:n't)?\(\))|mul\((\d+),(\d+)\)").unwrap();
    
    let mut enable = true;
    let res = pattern
        .captures_iter(file)
        .map(|cap| {
            match cap.get(1).map(|m| m.as_str()) {
                Some("do()") => {
                    enable = true;
                    0
                }
                Some("don't()") => {
                    enable = false;
                    0
                }
                _ => if enable {
                    let a = cap[2].parse::<i32>().unwrap();
                    let b = cap[3].parse::<i32>().unwrap();
                    a * b
                } else { 0 },      
            }
        })
        .sum::<i32>();
            
    println!("{}", res);
}
    

fn main() {
    let mut file = String::new();
    loop {
        let mut line = String::new();
        match stdin().read_line(&mut line) {
            Ok(0) => break,
            Ok(_) => file += &line,
            Err(_) => break,
        }
    }
    part1(&file);
    part2(&file);
}
